include/domain-types.h

```diff
@@ -121,6 +121,28 @@ struct Tag
 
 typedef std::vector<Tag> Tags;
 
+struct Payload
+{
+    Payload(const std::string& fileName)
+        : fileName(fileName)
+        , data(nullptr)
+        , dataSize(0)
+    {
+    }
+
+    Payload(const void* data, size_t dataSize, const std::string& mimeType)
+        : data(data)
+        , dataSize(dataSize)
+        , mimeType(mimeType)
+    {
+    }
+
+    std::string fileName;
+    const void* data;
```

Comment:
There are several points here:
1. Caller doesn't always have a vector. It may just have buffer he/she doesn't owne provided as pointer plus size. My approach covers this case too, while allowing input data to reside in a vector.
2. When method/function signature is const void* buf, size_t bufSize the ownership transfer is described in comments, and usually it isn't transferred as it incurs memory/performance penalty. Will providing comment for data and dataSize field close the discussion?
3. I totally disagree about using shared_array as described. I think shared_ptr/array is abused most of the time, as shared means that two or more owners that may (but not necessarily do) work with shared data simultaneously. Otherwise one would use unique_ptr. With shared one adds more confusion as it hides actual ownership transfer or its absence. 
4. Why do you suggest smart pointer here vs const void*? What problem does it solve?
5. I try to avoid complex data types in an interface (std::string being an exception). I don't understand a desire to pass data structures with allocators through module boundary (as an alternative consider using pure C interface as libcurl has) to then deal with problems such as those you described with Windows dlls. Note, that dealing with problems means ugly solutions that are arguably better, than C-style approach.